wtxmgr/db.go

```diff
@@ -0,0 +1,1395 @@
+	rootMinedBalance = []byte("bal")
+)
+
+// The root bucket's mined balance k/v pair records the total balance for all
+// unspent credits from mined transactions.  This includes immature outputs, and
+// outputs spent by mempool transactions, which must be considered when
+// returning the actual balance for a given number of block confirmations.  The
+// value is the amount serialized as a uint64.
+
+func fetchMinedBalance(ns walletdb.Bucket) (btcutil.Amount, error) {
+	v := ns.Get(rootMinedBalance)
+	if len(v) != 8 {
+		str := fmt.Sprintf("balance: short read (expected 8 bytes, "+
+			"read %v)", len(v))
+		return 0, storeError(ErrData, str, nil)
+	}
+	return btcutil.Amount(byteOrder.Uint64(v)), nil
+}
+
+func putMinedBalance(ns walletdb.Bucket, amt btcutil.Amount) error {
+	v := make([]byte, 8)
+	byteOrder.PutUint64(v, uint64(amt))
+	err := ns.Put(rootMinedBalance, v)
+	if err != nil {
+		str := "failed to put balance"
+		return storeError(ErrDatabase, str, err)
+	}
+	return nil
+}
+
+// Several data structures are given canonical serialization formats as either
+// keys or values.  These common formats allow keys and values to be reused
+// across different buckets.
+//
+// The canonical outpoint serialization format is:
+//
+//   [0:32]  Trasaction hash (32 bytes)
+//   [32:36] Output index (4 bytes)
+//
+// The canonical transaction hash serialization is simply the hash.
+
+func canonicalOutPoint(txHash *wire.ShaHash, index uint32) []byte {
+	k := make([]byte, 36)
+	copy(k, txHash[:])
+	byteOrder.PutUint32(k[32:36], index)
+	return k
+}
+
+func readCanonicalOutPoint(k []byte, op *wire.OutPoint) error {
+	if len(k) < 36 {
+		str := "short canonical outpoint"
+		return storeError(ErrData, str, nil)
+	}
+	copy(op.Hash[:], k)
+	op.Index = byteOrder.Uint32(k[32:36])
+	return nil
+}
+
+// Details regarding blocks are saved as k/v pairs in the blocks bucket.
+// blockRecords are keyed by their height.  The value is serialized as such:
+//
+//   [0:32]  Hash (32 bytes)
+//   [32:40] Unix time (8 bytes)
+//   [40:44] Number of transaction hashes (4 bytes)
+//   [44:]   For each transaction hash:
+//             Hash (32 bytes)
+
+func keyBlockRecord(height int32) []byte {
+	k := make([]byte, 4)
+	byteOrder.PutUint32(k, uint32(height))
+	return k
+}
+
+func valueBlockRecord(block *BlockMeta, txHash *wire.ShaHash) []byte {
+	v := make([]byte, 76)
+	copy(v, block.Hash[:])
+	byteOrder.PutUint64(v[32:40], uint64(block.Time.Unix()))
+	byteOrder.PutUint32(v[40:44], 1)
+	copy(v[44:76], txHash[:])
+	return v
+}
+
+// appendRawBlockRecord returns a new block record value with a transaction
+// hash appended to the end and an incremented number of transactions.
+func appendRawBlockRecord(v []byte, txHash *wire.ShaHash) ([]byte, error) {
+	if len(v) < 44 {
+		str := fmt.Sprintf("%s: short read (expected %d bytes, read %d)",
+			bucketBlocks, 44, len(v))
+		return nil, storeError(ErrData, str, nil)
+	}
+	newv := append(v[:len(v):len(v)], txHash[:]...)
```

Comment:
It would not be guaranteed to be new in that case.

http://play.golang.org/p/U9Xe_WHNmF